Local veneur server: Add timeouts for forwarding, make events flushes parallel

asf-stripe opened the pull request:
<!-- Checklist For PRs

* Ensure you've written tests where applicable
* Add a CHANGELOG.md entry describing your change, thank yourself!
* Document any changes to metrics or configuration
-->


#### Summary
This PR adds config options `forward_timeout`, `metrics_flush_timeout` and `span_flush_timeout`, controlling timeouts for each of these operations. This should help veneur avoid scenarios where an inability to quickly send metric batches (either upstream or to ~any of its storage backends) causes it to accumulate spans/metrics in memory and OOM the machine.

#### Motivation
Setting timeouts is good!


#### Test plan
Added regression tests, but we should also test this in a live system to ensure it actually errors in these cases.


#### Rollout/monitoring/revert plan
Merge & run!

asf-stripe commented:
R, @aditya-stripe?

ChimeraCoder commented:
What is the (live, not unit) test plan for this? Are we going to test it before merging/have we already?

asf-stripe commented:
Good question - I think to see how this works live, we can do it in two prongs:

* positive case / nothing fails under normal circumstances: Roll this to a host and watch it correctly submit metrics to our upstreams
* negative case: iptables some stuff on a throwaway host.

For the negative case, I don't think we can safely have proxies sleep (without throwing away all metrics from a host), and even if we could, that wouldn't exercise the direct flush mechanics /:

ChimeraCoder commented:
I'd feel most comfortable testing this on QA (both the positive and the negative cases) before merging.

Assigning back to you to fix the merge conflicts and do the testing; I'll take an additional look after that. :)

asf-stripe commented:
OK, phew, I finally pulled this forward. Once tests pass we can attempt a test.

asf-stripe commented:
I've found a reasonable way to test this & will try to do it next week:

* Roll out a configuration file with a broken upstream address to hosts that I'm testing this on: Since we run with a proxy in EC2 with strict security groups anyway, we can adjust the port. That should ensure all posts from that veneur server directly should time out.
* Check the graphs/alerting for those timeouts:
  * is it on a dashboard
  * is there a pager that goes off

I think we'll only need to roll this to a test box and to a global veneur box to see an effect.

asf-stripe commented:
OK, I've ported this forward. Been a few significant changes; PTAL, @aditya-stripe?

asf-stripe commented on example.yaml:
```diff
@@ -162,6 +162,28 @@ trace_max_length_bytes: 16384
 
+# How many metrics to include in the body of each POST to Datadog. Veneur
+# will post multiple times in parallel if the limit is exceeded.
+flush_max_per_body: 25000
```
how did this get added in this PR, but the FlushMaxPerBody above in config.go didn't? IDGI.

asf-stripe commented:
@aditya-stripe, I rebased this to the dev version again and fixed two problems I noticed (that outdated comment on example.yaml & a slow test) - mind taking another look?

ChimeraCoder commented:
I believe this is no longer necessary, but please reopen if I'm mistaken
